### DISCUSSION TITLE
RFC: `stacks`

### DISCUSSION THREAD

**Discussion by nksaraf** [nksaraf/vinxi#65] [Ideas]
Now that the base implementation is getting stable and proving itself to be quite flexible, I want to figure out the UX and composition story. Right now you need an app.config.js to do anything server related. This makes it difficult to enable an experience I really wish to enable. A powerful CLI that can compose functionality from NPM on the fly to build up an app config. 

I want to introduce the concept of `stacks`. They are the equivalent of `Vite` plugins but for the app itself, so more wide-reaching. In `vinxi`, `vite` plugins are applied to routers, since each one has its own dev/build pipelines. `Stacks` are the equivalent extension point for the app itself. 

With the ability to hook into the system using the `hookable` hooks, we can create a really composable setup where functionality can be bundled up and comsumed easily across the ecosystem. Whatever you can do with stacks, you will be able to do with the hooks themselves too. Maybe we can keep the naming the same too.

So, what is a stack?

1. It can add routers to the app
2. It can modify existing routers, changing settings, or adding vite plugins
3. More to come.. but basically hook into any step in the lifecycle and mutate stuff.

**Comment by kylegl** [comment]
Hi @nksaraf,

Quick question for you regarding these lifecycle hooks. I am using solidStart and would like to generate a sitemap during the build. I've looked through the vinxi code and can't quite figure out where exactly I could add something like that with solidStart. Ideally it would be after the router is built. Whats the best way to add hooks?

**Reply by nksaraf** [reply]
When you do createApp in app.conffig.ts you get an app reference back, you can register hooks on that.
